wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,346 @@
+from __future__ import annotations
+from typing import Iterator, List, Tuple, Optional, Any, BinaryIO
+
+from .at import at
+from .casts import to_clvm_object, int_from_bytes, int_to_bytes
+from .chia_dialect import CHIA_DIALECT
+from .clvm_rs import run_serialized_chia_program
+from .clvm_storage import CLVMStorage
+from .clvm_tree import CLVMTree
+from .curry_and_treehash import CurryTreehasher
+from .eval_error import EvalError
+from .replace import replace
+from .ser import sexp_from_stream, sexp_to_stream, sexp_to_bytes
+from .tree_hash import sha256_treehash
+
+
+MAX_COST = 0x7FFFFFFFFFFFFFFF
+
+
+class Program(CLVMStorage):
+    """
+    A wrapper around `CLVMStorage` providing many convenience functions.
+    """
+
+    curry_treehasher: CurryTreehasher = CurryTreehasher(CHIA_DIALECT)
+    _cached_serialization: Optional[bytes]
+
+    # serialization/deserialization
+
+    @classmethod
+    def parse(cls, f: BinaryIO) -> Program:
+        return sexp_from_stream(f, cls.new_pair, cls.new_atom)
+
+    def stream(self, f: BinaryIO) -> None:
+        sexp_to_stream(self, f)
+
+    @classmethod
+    def from_bytes(cls, blob: bytes, calculate_tree_hash: bool = True) -> Program:
+        obj, cursor = cls.from_bytes_with_cursor(
+            blob, 0, calculate_tree_hash=calculate_tree_hash
+        )
+        return obj
+
+    @classmethod
+    def from_bytes_with_cursor(
+        cls, blob: bytes, cursor: int, calculate_tree_hash: bool = True
+    ) -> Tuple[Program, int]:
+        tree = CLVMTree.from_bytes(
+            blob[cursor:], calculate_tree_hash=calculate_tree_hash
+        )
+        obj = cls.wrap(tree)
+        new_cursor = len(bytes(tree)) + cursor
+        return obj, new_cursor
+
+    @classmethod
+    def fromhex(cls, hexstr: str) -> Program:
+        return cls.from_bytes(bytes.fromhex(hexstr))
+
+    def __bytes__(self) -> bytes:
+        if self._cached_serialization is None:
+            self._cached_serialization = sexp_to_bytes(self)
```

Comment:
I think I'm misunderstanding something. Aren't the serialized form available in `self.blob`? You just need to pick the right slice from it. This looks like it's actually producing a new serialization.

Reply:
The `sexp_to_bytes` knows about the optional `_cached_serialization` memoizing. That's necessary because you might go `(q . MOD)` where the serialization for `MOD` is cached, so you want to check at each step down the tree and not just at the top.

I have a separate (but not checked in) benchmark tool which runs a bunch of cases and makes sure only the things that are supposed to be slow are slow, and everything else that should use memoized values (tree hashing, serialization, etc.) must run fast (less than .01 s). It's been useful in spotting performance bugs. I'm not sure how to check it in yet though, since benchmarking of this sort is so dependent on the platform it runs on (ie. how to generalize the concept of .01 s of time on my M2 Mac to "slow" on other less predictable platforms like in CI).